[registry] Bootstrap cluster with registry module enabled

vadimartynov opened the pull request:
## Description

This pull request adds the capability to:
1. Bootstrap a cluster with the registry module already enabled and operational
2. Use `moduleConfig/deckhouse` for registry settings instead of `initConfiguration`.

---

### Architecture of Bootstrap Cluster with Registry Module Enabled

**Bootstrap process (dhctl)**:

```mermaid
flowchart TD
    %% Configuration
    A[InitConfig]
    B[ModuleConfig Deckhouse]
    
    %% Infrastructure
    subgraph Metaconfig [Metaconfig parsing]
        C[Registry config parsing]
    end
    D[Bootstrap stages ...]
    
    %% Bashible
    subgraph Bashible [Bashible]
        E[Create PKI secret]
    end
    
    %% Deckhouse Installation
    subgraph Installation [Deckhouse Installation]
        direction TB
        F[Get PKI, create manifests]
        G[Install, run and verify Deckhouse]
        H[Run and verify registry module]
        I[Delete PKI secret]
    end
    
    %% Completion
    J[Other bootstrap stages ...]
    
    %% Flow structure
    A --> C
    B --> C
    Metaconfig --> D
    D --> Bashible
    Bashible --> Installation
    F --> G
    G --> H
    H --> I
    Installation --> J
```

**Bootstrap process (module registry)**:

```mermaid
flowchart TD
    %% Main process
    subgraph Reconciliation [Registry module reconciliation]
        direction TB
        B[Get PKI secret on first run]
        C[Set annotation on PKI]
        D[Fill module conditions]
        
        B --> C
        C --> D
    end
    
    %% Results
    E[PKI data obtained]
    F[Registry module ready]
    
    %% Flow structure
    Reconciliation --> E
    Reconciliation --> F
```

---

### Registry Configuration During Bootstrap

To enable the registry module operation during cluster bootstrap, support for parsing data from `moduleConfig/deckhouse` has been added. Configuration Options:

**initConfiguration** (Legacy Configuration):
- Only supports Unmanaged mode without registry module initialization
- Maintained for backward compatibility

**moduleConfig/deckhouse**:
- Supports **both Direct and Unmanaged modes**
- Mode is determined by user configuration

**default**:
- Uses **Direct mode** if supported, otherwise **Unmanaged mode**
- Populated with Community Edition default values

> [!IMPORTANT]  
> - **Conflict Prevention**: To avoid configuration conflicts between registry settings in InitConfig and deckhouse ModuleConfig, registry parameters are now optional in `initConfiguration`.
> - **Conflict Handling**: If both configurations are specified, an error will be displayed indicating a configuration conflict.

**Previous Configuration:**

```yaml
apiVersion: deckhouse.io/v1
kind: InitConfiguration
deckhouse:
  imagesRepo: test-registry.io/some/path
  registryScheme: HTTPS
  registryDockerCfg: <base64 docker cfg>
  registryCA: <CA>
```

**New Configuration Options:**

```yaml
# Legacy configuration still supported - Unmanaged mode without registry module initialization:
apiVersion: deckhouse.io/v1
kind: InitConfiguration
deckhouse:
  imagesRepo: test-registry.io/some/path
  registryScheme: HTTPS
  registryDockerCfg: <base64 docker cfg>
  registryCA: <CA>
---
# New configuration - Unmanaged mode
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: deckhouse
spec:
  version: 1
  enabled: true
  settings:
    registry:
      mode: Unmanaged
      unmanaged:
        imagesRepo: test-registry.io/some/path
        scheme: HTTPS
        username: <username>
        password: <password>
        ca: <CA>
---
# New configuration - Direct mode
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  name: deckhouse
spec:
  version: 1
  enabled: true
  settings:
    registry:
      mode: Direct
      direct:
        imagesRepo: test-registry.io/some/path
        scheme: HTTPS
        username: <username>
        password: <password>
        ca: <CA>
```

---

## Why do we need it, and what problem does it solve?

- We need to bootstrap clusters with the registry module immediately enabled and functional
- We must be able to configure registry through a unified interface (`moduleConfig/deckhouse` instead of `initConfig`) both during bootstrap and in a running cluster


## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
<!---
  Describe the changes so they will be included in a release changelog.

  Find examples and documentation below, or visit the [Guidelines for working with PRs](https://github.com/deckhouse/deckhouse/wiki/Guidelines-for-working-with-PRs).
-->

```changes
section: dhctl
type: feature
summary: Added bootstrap support with the registry module for Direct and Unmanaged modes.
impact_level: default
---
section: candi
type: fix
summary: Moved the default values for registry in initConfiguration to dhctl.
impact_level: default
---
section: registry
type: feature
summary: Added bootstrap support for Direct and Unmanaged modes.
impact_level: default
---
section: node-manager
type: fix
summary: Updated go dependencies in the bashible-api-server.
impact_level: default
```

<!---

`impact_level: default` adds to changelog as usual, this is the default that can be omitted
`impact_level: high`    something important for users, the impact will be copied to "Know Before Update" section
`impact_level: low`     omitted in changelog YAML; note there is `type:chore` for chores

Tip for the section field:

  - <kebab-case of a module>, e.g. "cloud-provider-aws", "node-manager"
  - "ci", has forced low impact
  - "docs", includes website changes, should have low impact
  - "candi"
  - "deckhouse-controller"
  - "dhctl"
  - "global-hooks"
  - "go_lib"
  - "helm_lib"
  - "jq_lib"
  - "shell_lib"
  - "testing", has forced low impact
  - "tools", has forced low impact

Find changed sections:

gh pr diff   $PULL_REQUEST_NUMBER   |
  egrep "^([+]{3} b|[-]{3} a)/" |
  cut -d/ -f2- |
  sed 's#^ee/##' |
  sed 's#^fe/##' |
  sed 's#^modules/##' |
  sed 's#[0-9][0-9][0-9]-##' |
  egrep -v 'Makefile' |       # add file exclusion here
  cut -d/ -f1 |
  sort |
  uniq

Find all possible sections (excluding ci):

node -e 'console.log(require("./.github/scripts/js/changelog-find-sections.js")().join("\n"))'
-->

vadimartynov committed:
Mooved docker cfg helpers to go_lib/registry

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Mooved registry to own package

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Implementation of the bootstrap registry phase

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Dhctl: fixed registry bootstrap secret creation

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Module: Added getting a bootstrap secret

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Make generate

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed initSecret processing

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Renamed bootstrap secret to init secret

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Module: initialized state with init config

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Minor fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Added registry bashible config secret initialization

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Mooved ComputeHash to go_lib/registry

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Bump bashible-api-server

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed dhctl errors

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed dhctl

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed dhctl

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed dhctl

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed dhctl

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Module: fixed initSecret process

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Dhclt: Added license field in deckhouse mc spec

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Dhctl:
- Minor improvements;
- Fixed the build of bashible models

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Dhctl:
- fixed registry obj init;
- fixed mc deckhouse filling

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed panic

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Module: fixed pki state, removed incorrect users

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed registry-pki state secret

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Dhclt: fixed bashible config secret

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Dhclt: used deckhouse registry secret model from go_lib/registry

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Golang ci lint

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Log improvement

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
Module: added check in bashible transition stage

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Module: small code improvement

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Module: fixed error

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed moduleSource access error:
- added hostAliases for deckhouse deployment;
- added hostNetwork for incluster proxy;

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
Update dhctl:
- registry refactoring;
- initialized pki provider;

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] Added optional data. If the registry module is disabled

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] InitSecret is no longer deleted by the module

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Moved the creation of the init secret to bassible

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed d8-system namespace update:
- [dhctl] Added update func in CreateDeckhouseManifests
- [bashible] Added d8-system namespace creation in registry bootstrap step

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed cri validation

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed kubeadm and bashible tpl

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
Rename Builder to ConfigBuilder

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Improved work with init secret

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed PKI:
- added ClusterPKIManager;
- merged init and pki structs;

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] Improved PKI operation and added the WaitForRegistryInitialization step

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] Error msg fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] Error msg fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Updated WaitForRegistryInitialization func

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] refactoring

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Improved registry pkg

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhclt] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhclt] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhclt] Improved registry pkg

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Preflight checks are unblocked

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Base tests are unblocked

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Meta tests are unblocked

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Deckhouse install tests are unblocked

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Bootstrap tests are unblocked

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Added test config builder and fixed tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] Fixed base tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Added meta tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed meta tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed meta tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed base and meta tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Fixed:
- Returned the old creation of the registry-pki secret;
- Returned the old user model;
- Moved the creation of the PKI to the beginning of transitionToDirect;
- Removed unused users from the init secret;

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Removed unused funcs and methods

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Mooved NewTestConfig to a separate file

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Replaced mode interface with modeSettings and modeModel, refactoring

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed errors

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] Added deckhouse config test

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Fixed pki secret for unmanaged mode

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] added helpers test

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Removed default registry settings from initConfig

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed after rebase

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Added tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov commented:
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.32 19747275993 19747275993-1-con-1-32 /sys/deckhouse-oss/install:pr16103-ce redos@158.160.43.212

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

deckhouse-BOaTswain commented:
Aye, aye, @vadimartynov. I've started the workflow for label 'e2e/run/eks'...
<!-- workflow_start -->
  :fast_forward: `e2e: EKS, Containerd, Kubernetes 1.32` for `deckhouse:registry-bootstrap` [started](https://github.com/deckhouse/deckhouse/actions/runs/19758424436).

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[dhctl] Fixed tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Mooved SplitAddressAndPath to helm lib

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Removed Direct and Unmanaged settings

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed typo

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Mooved const to go_lib

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhclt] Mooved types to go_lib

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Simplified PKI

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Made functions private

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed mode model

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Mooved ModuleEnabledCRI to go_lib

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Mooved registry config to a separate method

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov commented on dhctl/pkg/config/registry/types/const.go:
```diff
@@ -0,0 +1,41 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
```
Done

vadimartynov commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,207 @@
+	Remote types.Data
+}
+
+func NewModeSettings(settings types.DeckhouseSettings) (ModeSettings, error) {
```
Done

vadimartynov commented on dhctl/pkg/config/meta.go:
```diff
@@ -88,17 +90,69 @@ func (m *MetaConfig) Prepare(ctx context.Context, preparatorProvider MetaConfigP
+	}
+
+	// Prepare registry configuration
+	{
```
Done

vadimartynov committed:
[bashible-api-server] go mod tidy

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Moved all methods under struct definition

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Config small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[go_lib] Changed Correct to ApplySettings

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Added processInit

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] init fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Fixed incluster-proxy tolerations

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Small improvement

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov commented on modules/038-registry/hooks/orchestrator/hook.go:
```diff
@@ -244,6 +292,14 @@ func handle(ctx context.Context, input *go_hook.HookInput) error {
+	if initSecret.IsExist && !initSecret.Applied {
+		err = values.State.processInit(input.Logger, inputs)
+		if err != nil {
+			return fmt.Errorf("cannot process: %w", err)
```
Done

vadimartynov committed:
[dhctl] Added a configuration pre-check to determine if the module will start

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Renamed NewTestConfig -> TestConfigBuilder

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Default direct mode

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Improved work with basible and init models

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Added tests for initConfig

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Small improvement

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov commented on candi/openapi/init_configuration.yaml:
```diff
@@ -40,15 +40,13 @@ apiVersions:
 
               For example, to access the container registry `registry.company.my` under the user `user` with the password `P@ssw0rd` it will be `eyJhdXRocyI6eyJyZWdpc3RyeS5jb21wYW55Lm15Ijp7ImF1dGgiOiJkWE5sY2pwUVFITnpkekJ5WkFvPSJ9fX0K` (string `{"auths":{"registry.company.my":{"auth":"dXNlcjpQQHNzdzByZAo="}}}` in Base64).
             minLength: 1
-            default: eyJhdXRocyI6IHsgInJlZ2lzdHJ5LmRlY2tob3VzZS5pbyI6IHt9fX0=
```
Added tests. Done

vadimartynov committed:
[dhctl] Added mode and manifests tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

deckhouse-BOaTswain commented:
Aye, aye, @vadimartynov. I've started the workflow for label 'e2e/run/eks'...
<!-- workflow_start -->
  :fast_forward: `e2e: EKS, Containerd, Kubernetes 1.32` for `deckhouse:registry-bootstrap` [started](https://github.com/deckhouse/deckhouse/actions/runs/19937045056).

vadimartynov committed:
[dhctl] Fixed default mode

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[module] Added initialization of unmanaged mode params

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

deckhouse-BOaTswain commented:
Aye, aye, @vadimartynov. I've started the workflow for label 'e2e/run/yandex-cloud'...
<!-- workflow_start -->
  :fast_forward: `e2e: Yandex.Cloud, Containerd, Kubernetes 1.32` for `deckhouse:registry-bootstrap` [started](https://github.com/deckhouse/deckhouse/actions/runs/20053273225).

vadimartynov commented:
/e2e/abort eks;WithoutNAT;containerd;1.32 20053273632 20053273632-1-con-1-32 /sys/deckhouse-oss/install:pr16103-ce

vadimartynov committed:
[dhctl] ModuleEnable -> LegacyMode

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

RaveNoX committed:
refactoring

Signed-off-by: Artur Kraev <artur.kraev@flant.com>

RaveNoX commented on dhctl/pkg/config/registry/manifests.go:
```diff
@@ -0,0 +1,115 @@
+// =======================
+// Secrets
+// =======================
+func (b *ManifestBuilder) DeckhouseRegistrySecretData(getPKI func() (PKI, error)) (map[string][]byte, error) {
```
it will be nice to introduce type like type SecretData map[string][]byte and use it as return type

RaveNoX commented on dhctl/pkg/config/registry/manifests.go:
```diff
@@ -0,0 +1,115 @@
+	}
+}
+
+func (b *ManifestBuilder) BashibleTplCtx(getPKI func() (PKI, error)) (map[string]interface{}, error) {
```
it looks like `getPKI func() (PKI, error)` is good candidate for type

RaveNoX commented on dhctl/pkg/config/registry/manifests.go:
```diff
@@ -0,0 +1,115 @@
+func (b *ManifestBuilder) BashibleTplCtx(getPKI func() (PKI, error)) (map[string]interface{}, error) {
+	cfg, err := b.modeModel.BashibleConfig()
+	if err != nil {
+		return nil, err
```
please wrap err

vadimartynov replied:
Done

RaveNoX commented on dhctl/pkg/config/registry/manifests.go:
```diff
@@ -0,0 +1,115 @@
+
+	mapCtx, err := ctx.ToMap()
+	if err != nil {
+		return nil, err
```
please wrap err

RaveNoX commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,179 @@
+			Remote: remote,
+		}, nil
+	}
+	return ModeSettings{}, ErrUnknownMode
```
move this to `default` case

RaveNoX commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,179 @@
+	case settings.Direct != nil:
+		remote := Data{}
+		remote.FromRegistrySettings(*settings.Direct)
+		return ModeSettings{
```
please add empty like before return

vadimartynov replied:
Done

RaveNoX commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,179 @@
+	case settings.Unmanaged != nil:
+		remote := Data{}
+		remote.FromRegistrySettings(*settings.Unmanaged)
+		return ModeSettings{
```
please add like before return

RaveNoX commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,179 @@
+	case constant.ModeUnmanaged:
+		return s.unmanagedModel()
+	}
+	panic(ErrUnknownMode)
```
move it to default case

RaveNoX commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,179 @@
+	panic(ErrUnknownMode)
+}
+
+func (s ModeSettings) directModel() ModeModel {
```
toDirectModel

RaveNoX commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,179 @@
+	case constant.ModeUnmanaged:
+		return m.RemoteData, nil
+	}
+	return Data{}, ErrUnknownMode
```
move it to default case

RaveNoX committed:
refactoring

Signed-off-by: Artur Kraev <artur.kraev@flant.com>

RaveNoX commented on dhctl/pkg/config/registry/steps.go:
```diff
@@ -0,0 +1,179 @@
+	return conditions, nil
+}
+
+func getInitSecretStatus(ctx context.Context, kubeClient client.KubeClient) (bool, bool, error) {
```
please clarify return values in func comment

vadimartynov committed:
[dhctl] Fixed and added tests of modes and manifests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

RaveNoX commented on dhctl/pkg/config/meta.go:
```diff
@@ -134,6 +138,97 @@ func (m *MetaConfig) Prepare(ctx context.Context, preparatorProvider MetaConfigP
+	// Extract configuration from initConfig
+	if m.DeckhouseConfig.ImagesRepo != "" ||
+		m.DeckhouseConfig.RegistryDockerCfg != "" ||
+		m.DeckhouseConfig.RegistryCA != "" {
```
it looks like this comparsion chain should be wrapped to the method in `DeckhouseConfig`

vadimartynov committed:
[dhctl] Fixed naming

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

RaveNoX commented on dhctl/pkg/kubernetes/actions/deckhouse/install.go:
```diff
@@ -196,7 +197,8 @@ func CreateDeckhouseManifests(
 			},
 			UpdateFunc: func(manifest interface{}) error {
-				return nil
+				_, err := kubeCl.CoreV1().Namespaces().Update(ctx, manifest.(*apiv1.Namespace), metav1.UpdateOptions{})
```
please move each call in method chain to the separate line

RaveNoX commented on dhctl/pkg/kubernetes/actions/deckhouse/install.go:
```diff
@@ -288,10 +290,48 @@ func CreateDeckhouseManifests(
+		Name:     `Secret "deckhouse-registry"`,
+		Manifest: func() interface{} { return manifests.DeckhouseRegistrySecret(deckhouseRegistrySecretData) },
+		CreateFunc: func(manifest interface{}) error {
+			_, err := kubeCl.CoreV1().Secrets("d8-system").Get(ctx, manifest.(*apiv1.Secret).GetName(), metav1.GetOptions{})
```
please move each call in method chain to the separate line

RaveNoX commented on dhctl/pkg/kubernetes/actions/manifests/manifests.go:
```diff
@@ -555,23 +550,29 @@ func DeckhouseRegistrySecret(registry config.RegistryData) *apiv1.Secret {
 
-	if registry.CA != "" {
-		ret.Data["ca"] = []byte(registry.CA)
+func RegistryBashibleConfigSecret(data map[string][]byte) *apiv1.Secret {
```
`map[string][]byte` is good candidate for new type

RaveNoX commented on go_lib/registry/models/deckhouse-registry/config.go:
```diff
@@ -63,6 +63,16 @@ func (cfg *Config) ToBase64SecretData() map[string]string {
 	}
 }
 
+func (cfg *Config) ToMap() map[string][]byte {
```
it looks like `map[string][]byte` is good candidate for the new type

RaveNoX commented on go_lib/registry/models/init-secret/config.go:
```diff
@@ -0,0 +1,46 @@
+	CA *CertKey `json:"ca,omitempty" yaml:"ca,omitempty"`
+}
+
+func (c Config) ToMap() map[string]interface{} {
```
`interface{}` -> `any`

RaveNoX commented on go_lib/registry/models/init-config/config.go:
```diff
@@ -0,0 +1,108 @@
+	if err != nil {
+		return module_config.RegistrySettings{}, fmt.Errorf("failed to decode registryDockerCfg: %w", err)
+	}
+	username, password, err := helpers.CredsFromDockerCfg(dockerCfgDecode, address)
```
please add blank line above

RaveNoX commented on go_lib/registry/models/init-config/config.go:
```diff
@@ -0,0 +1,108 @@
+	}
+	registrySettings.Username = username
+	registrySettings.Password = password
+	return registrySettings, nil
```
please add blank line above

RaveNoX commented on go_lib/registry/models/module-config/deckhouse.go:
```diff
@@ -0,0 +1,173 @@
+	}
+	if userSettings.CheckMode != "" {
+		settings.CheckMode = userSettings.CheckMode
+	}
```
please add blank line before each `if`

vadimartynov replied:
Done

RaveNoX commented on modules/038-registry/hooks/orchestrator/hook.go:
```diff
@@ -260,10 +316,25 @@ func handle(ctx context.Context, input *go_hook.HookInput) error {
+		}
+		input.PatchCollector.PatchWithMerge(
+			patch, "v1", "Secret", "d8-system", "registry-init")
+	}
```
`interface{}` -> `any`

vadimartynov replied:
Done

vadimartynov committed:
[dhctl] Added tests for manifests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
interface -> any

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fmt

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed:
- added secretData contextData and getPKI types;
- wrap errors;
- added names to the returned tuples

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Remote -> RemoteData

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
added "to" prefix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov commented on dhctl/pkg/config/registry/config.go:
```diff
@@ -0,0 +1,119 @@
+	return newManifestBuilder(c.Settings.ToModel(), c.LegacyMode)
+}
+
+func (c *Config) DeckhouseSettingsToMap() (bool, map[string]any, error) {
```
Done

vadimartynov commented on dhctl/pkg/config/registry/manifests.go:
```diff
@@ -0,0 +1,115 @@
+	}
+}
+
+func (b *ManifestBuilder) BashibleTplCtx(getPKI func() (PKI, error)) (map[string]interface{}, error) {
```
Done

vadimartynov commented on dhctl/pkg/config/registry/manifests.go:
```diff
@@ -0,0 +1,115 @@
+// =======================
+// Secrets
+// =======================
+func (b *ManifestBuilder) DeckhouseRegistrySecretData(getPKI func() (PKI, error)) (map[string][]byte, error) {
```
Done

vadimartynov commented on dhctl/pkg/config/registry/mode.go:
```diff
@@ -0,0 +1,179 @@
+
+	return map[string]bashible.ConfigHosts{
+		host: {
+			Mirrors: []bashible.ConfigMirrorHost{cfgMirror}},
```
Done

vadimartynov commented on modules/038-registry/hooks/orchestrator/state.go:
```diff
@@ -100,6 +100,65 @@ func (state *State) clearConditions() {
+	// Set Bashible ActualParams
+	var bashibleActualParams *bashible.ModeParams
+	var bashibleUnmanagedParams *bashible.UnmanagedModeParams
+	switch inputs.Params.Mode {
```
Done

vadimartynov committed:
Minor fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov commented on dhctl/go.mod:
```diff
@@ -115,6 +116,7 @@ require (
 	github.com/imdario/mergo v0.3.16 // indirect
 	github.com/inconshreveable/mousetrap v1.1.0 // indirect
 	github.com/jellydator/ttlcache/v3 v3.4.0 // indirect
+	github.com/jmoiron/sqlx v1.3.5 // indirect
```
```
github.com/deckhouse/deckhouse/go_lib/registry github.com/jmoiron/sqlx@v1.3.5
github.com/cloudflare/cfssl@v1.6.5 github.com/jmoiron/sqlx@v1.3.5
github.com/jmoiron/sqlx@v1.3.5 github.com/go-sql-driver/mysql@v1.6.0
github.com/jmoiron/sqlx@v1.3.5 github.com/lib/pq@v1.2.0
github.com/jmoiron/sqlx@v1.3.5 github.com/mattn/go-sqlite3@v1.14.6
```

vadimartynov committed:
moved defaults to a separate file

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Fixed types:
- make new type instead of type alias in registry const
- fixed registry config aliases

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Mooved config_for_testing.go to registry mocks

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Remove named return vars, add comments instead

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Mooved types to types.go

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Added init field to bashible ctx model

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Mooved pki types to types

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Now manifests.BashibleContext returns the model

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Removed contextData, now ManifestBuilder returns structs with the toMap method

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixes and improvements:
- Removed DeckhouseSettingsToMap
- SecretData is no longer aliases
- Improved the toMap method for DeckhouseSettings

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Added and fixed comments

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Added some tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Improved meta.prepareRegistry

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[candi] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov commented on dhctl/pkg/config/meta.go:
```diff
@@ -134,6 +138,97 @@ func (m *MetaConfig) Prepare(ctx context.Context, preparatorProvider MetaConfigP
+			return fmt.Errorf("get registry settings from 'moduleConfig/deckhouse': %w", err)
+		}
+		deckhouseSettings = &decoded
+		break
```
Done

vadimartynov committed:
[dhctl] Fixed tests

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
[candi] Removed x-doc-default

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

RaveNoX commented on dhctl/pkg/config/meta.go:
```diff
@@ -134,6 +136,91 @@ func (m *MetaConfig) Prepare(ctx context.Context, preparatorProvider MetaConfigP
+		if err := json.Unmarshal(raw, &ret); err != nil {
+			return nil, fmt.Errorf("unmarshal deckhouse settings: %w", err)
+		}
+		return &ret, nil
```
can we move this login from loop?

RaveNoX commented on dhctl/pkg/config/types.go:
```diff
@@ -74,6 +76,21 @@ type DeckhouseClusterConfig struct {
+	if config.ImagesRepo != "" ||
+		config.RegistryScheme != "" ||
+		config.RegistryDockerCfg != "" ||
+		config.RegistryCA != "" {
```
move this conditions to named method

vadimartynov replied:
Done

RaveNoX commented on go_lib/registry/const/defaults.go:
```diff
@@ -16,23 +16,9 @@ limitations under the License.
-	Relax   CheckModeType = "Relax"
+	UnknownVersion  = "unknown"
+	LicenseUsername = "license-token"
+	CEImagesRepo    = "registry.deckhouse.io/deckhouse/ce"
```
DefaultImagesRepo

vadimartynov replied:
Done

vadimartynov committed:
[dhctl] Fixed checkReady step

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Added hasRegistryInitConfig method in DeckhouseClusterConfig

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed MetaConfig.registryDeckhouseSettings

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Fixed field -> section

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Renamed CEImagesRepo and CEScheme to DefaultImagesRepo and DefaultScheme

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Renamed pkgs for go generate

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Added a hack to create deepCopy methods for models

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Added and applied deepCopy

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Fixed Copyright Validation

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Fixed go generate

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Mooved generate.go to hack

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Fixed: fallback to unmanaged when registry configs overwritten during bootstrap

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[go_lib] Manual implementation of deepCopy

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Fixed transitionToUnmanaged conditions

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[dhctl] Fixed logs info for wait func in bootstrap stage

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

vadimartynov committed:
Merge branch 'main' into registry-bootstrap

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

deckhouse-BOaTswain commented:
Aye, aye, @vadimartynov. I've started the workflow for label 'e2e/run/yandex-cloud'...
<!-- workflow_start -->
  :fast_forward: `e2e: Yandex.Cloud, Containerd, Kubernetes 1.32` for `deckhouse:registry-bootstrap` [started](https://github.com/deckhouse/deckhouse/actions/runs/20365929321).


Yandex.Cloud-WithoutNAT-Containerd-1.32 - Connection string: `ssh  redos@158.160.45.14`

deckhouse-BOaTswain commented:
:green_circle: `destroy cluster: Yandex.Cloud` for `refs/heads/main` [succeeded](https://github.com/deckhouse/deckhouse/actions/runs/20369638554) in 2m35s.
<details><summary>Workflow details</summary>


:green_circle: `destroy cluster: Yandex.Cloud, Containerd, Kubernetes 1.32` succeeded in 2m2s.<!-- result-for: destroy cluster: Yandex.Cloud, Containerd, Kubernetes 1.32 -->

</details>

vadimartynov committed:
[module] Freezed input params during initialization

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Small fix

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Fixed InitSecret snap get

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] Fixed init params state

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov committed:
[module] improved logs

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

RaveNoX committed:
refactor

Signed-off-by: Artur Kraev <artur.kraev@flant.com>

deckhouse-BOaTswain commented:
:green_circle: `destroy cluster: Yandex.Cloud` for `refs/heads/main` [succeeded](https://github.com/deckhouse/deckhouse/actions/runs/20456267270) in 3m2s.
<details><summary>Workflow details</summary>


:green_circle: `destroy cluster: Yandex.Cloud, Containerd, Kubernetes 1.32` succeeded in 2m3s.<!-- result-for: destroy cluster: Yandex.Cloud, Containerd, Kubernetes 1.32 -->

</details>

deckhouse-BOaTswain commented:
Aye, aye, @vadimartynov. I've started the workflow for label 'e2e/run/eks'...
<!-- workflow_start -->
  :fast_forward: `e2e: EKS, Containerd, Kubernetes 1.32` for `deckhouse:registry-bootstrap` [started](https://github.com/deckhouse/deckhouse/actions/runs/20456332334).

nabokihms commented on candi/openapi/init_configuration.yaml:
```diff
@@ -40,15 +40,15 @@ apiVersions:
             minLength: 1
-            default: eyJhdXRocyI6IHsgInJlZ2lzdHJ5LmRlY2tob3VzZS5pbyI6IHt9fX0=
+            x-examples:
+              - eyJhdXRocyI6IHsgInJlZ2lzdHJ5LmRlY2tob3VzZS5pbyI6IHt9fX0=
```
```suggestion
```

vadimartynov replied:
The `registryDockerCfg` parameter defaults to an empty string (`""`).
• If the value is empty → parsing is skipped.
• If a value is provided → it is validated and parsed.

Fixed `x-doc-default` in https://github.com/deckhouse/deckhouse/pull/16103/commits/320ce5b940c3253f1124f7ca15e82359927a3da6

vadimartynov committed:
[openapi] Added x-doc-deprecated and x-doc-default

Signed-off-by: vadimartynov <vadim.martynov@flant.com>

vadimartynov commented on candi/openapi/init_configuration.yaml:
```diff
@@ -40,15 +40,15 @@ apiVersions:
 
               The address matches the edition of Deckhouse used. The public container registry address for Deckhouse Enterprise Edition is `registry.deckhouse.io/deckhouse/ee`.
             minLength: 1
-            default: registry.deckhouse.io/deckhouse/ce
```
Fixed in https://github.com/deckhouse/deckhouse/pull/16103/commits/320ce5b940c3253f1124f7ca15e82359927a3da6

deckhouse-BOaTswain commented:
Aye, aye, @vadimartynov. I've started the workflow for label 'e2e/run/yandex-cloud'...
<!-- workflow_start -->
  :fast_forward: `e2e: Yandex.Cloud, Containerd, Kubernetes 1.32` for `deckhouse:registry-bootstrap` [started](https://github.com/deckhouse/deckhouse/actions/runs/20466198057).


Yandex.Cloud-WithoutNAT-Containerd-1.32 - Connection string: `ssh  redos@158.160.120.180`

vadimartynov committed:
[module] Fixed generation type

Signed-off-by: vadimartynov <vadim.martynov@flant.com>
